### PR TITLE
[Inductor][CPP] Enable the LocalBuffer

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126056
* __->__ #126055

**Summary**
Currently, the Inductor CPP backend [generated code](https://gist.github.com/leslie-fang-intel/98f91d43dabed581a1ffe23daf133a65#file-bf16-softmax-generated-code-wo-local-buffer-py) for `Softmax` with BF16 data type is significantly slower than the [ATen Implementation](https://github.com/pytorch/pytorch/blob/9a2beb862d9c30f037c9b2eac878ec3f9227a5e2/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L149). Upon comparing the generated code with ATen, the performance bottleneck appears to be related to the usage of [local buffer in ATen](https://github.com/pytorch/pytorch/blob/9a2beb862d9c30f037c9b2eac878ec3f9227a5e2/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L159-L160). 

In the current implementation, the Inductor uses the output buffer of Kernel Group Args to store and load temporary result (such as `exp`), since this buffer is corresponding to a `SchedulerNode`. Each thread accesses a portion of this output buffer via indexing. However, since this buffer (take this `exp` as example) is only utilized internally within decomposed `softmax`, this buffer can be replaced with a thread-local buffer similar to ATen's approach.

In this PR, we have introduced the optimizations of `LocalBuffer`. Following this enhancement, the [new generated Inductor code with local buffer](https://gist.github.com/leslie-fang-intel/98f91d43dabed581a1ffe23daf133a65#file-bf16-softmax-generated-code-w-local-buffer-py) for BF16 `Softmax` demonstrates significantly improved performance. Running the benchmark [here](https://gist.github.com/leslie-fang-intel/37d81441237b5139c8295f5e6c4cd31a) to test this BF16 `Softmax` case on an 8480 Xeon server shows similar performance between the Inductor CPP Backend and the ATen implementation.

**Implementation**
* In the codegen of `codegen_outer_loop_node`, we will analyze `codegen_outer_loop_node` to find which scheduler node can be replaced with local buffer. And we save this information into `CPPKernelProxy` and propagate to `CPPKernel`.
* In the codegen phase of `CPPKernel`, we will save/load this `LocalBuffer` instead of original buffer.
* In the codegen Loop phase, we will allocate the `LocalBuffer` with size and data type.

**Test Plan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_local_buffer_in_outer_loop_fusion
```

**Next Step**
In next PR, we will remove the output buffer which replaced to local buffer in the Args of Kernel Group and the allocation of this buffer in wrapper code.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -3605,48 +3718,162 @@ def can_fuse_vertical(self, node1, node2):
...
+                        user.node in node.get_nodes() for user in scheduler_node.users
+                    ):
+                        local_buffer = LocalBuffer(
+                            scheduler_node,
+                            keep_idx_id,
+                            call_ranges[keep_idx_id],
+                            scheduler_node.node.layout.get_dtype(),
+                        )
+                        local_buffers.append(local_buffer)
+                        # At most 1 node with local buf for each OuterLoopFusedSchedulerNode
```
Comment (peterbell10) [inline torch/_inductor/codegen/cpp.py:3782]: I'm curious why you restrict this to one local buffer? It seems fairly straight-forward to map `load("buf0", ...)` onto `local_buffers["buf0"]`.

Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:3782]: Thanks for the comment. For the typical case (softmax) I met, there is only one local buffer. Will think about how to extend this implementation.
